usr/lib/common/key.c

```diff
@@ -1660,6 +1663,133 @@ CK_RV rsa_priv_unwrap_get_data(TEMPLATE *tmpl,
+        TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_KEYFORM for the key.\n");
+        return CKR_TEMPLATE_INCOMPLETE;
+    }
+
+    /* Check if it's an expected keyform */
+    if (*(CK_ULONG *) keyform->pValue != IBM_DILITHIUM_KEYFORM_ROUND2) {
+        TRACE_ERROR("This key has an unexpected CKA_IBM_DILITHIUM_KEYFORM: %ld \n",
+                    *(CK_ULONG *) keyform->pValue);
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+
+    /* Check if key given as opaque blob */
+    if (template_attribute_find(tmpl, CKA_IBM_OPAQUE, &opaque) == FALSE) {
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_RHO, &rho) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_RHO for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_SEED, &seed) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_SEED for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_TR, &tr) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_TR for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_S1, &s1) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_S1 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_S2, &s2) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_S2 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T0, &t0) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_T0 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T1, &t1) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_T1 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+    }
+
+    rc = ber_encode_IBM_DilithiumPrivateKey(length_only, data, data_len,
+                                  rho, seed, tr, s1, s2, t0, t1,
+                                  opaque);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_encode_IBM_DilithiumPrivateKey failed\n");
+    }
+
+    return rc;
+}
+
+CK_RV ibm_dilithium_priv_unwrap_get_data(TEMPLATE *tmpl, CK_BYTE *data,
+                                         CK_ULONG total_length)
+{
+    CK_ATTRIBUTE *rho = NULL;
+    CK_ATTRIBUTE *t1 = NULL;
+    CK_RV rc;
+
+    rc = ber_decode_IBM_DilithiumPublicKey(data, total_length, &rho, &t1);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_decode_DilithiumPublicKey failed\n");
+        return rc;
+    }
+
+    rc = template_update_attribute(tmpl, rho);
+    if (rc != CKR_OK)
+        TRACE_DEVEL("template_update_attribute(CKA_IBM_DILITHIUM_RHO) failed\n");
```

Comment:
Shouldn't it return the return code in case of an error?

Reply:
yes, agreed. 
As all other ..._priv_unwrap_get_data routines are coded in the same way, we should also fix these. I wonder if someone in the past intentionally introduced this logic ....